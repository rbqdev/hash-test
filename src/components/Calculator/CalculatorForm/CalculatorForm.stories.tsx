import { Story, Meta } from "@storybook/react/types-6-0";
import CalculatorForm from ".";

export default {
  title: "CalculatorForm",
  component: CalculatorForm
} as Meta;

export const Default: Story = args => <CalculatorForm {...args} />;
