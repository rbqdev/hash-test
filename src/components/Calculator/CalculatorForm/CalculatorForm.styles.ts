import styled from "styled-components";
import { _constants as themeConstants } from "@styles/themes";
import media from "styled-media-query";

const { THEME_LIGHT } = themeConstants;

export const Wrapper = styled.section`
  width: 375px;
  padding: 41px 70px 44px 56px;
  border-radius: 4px 0 0 4px;
  background: ${({ theme: { title, colors } }) =>
    title === THEME_LIGHT ? colors.white : colors.secondaryBg};

  ${media.lessThan("medium")`
    width: 100%;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    padding: 25px;

    form {
      width: 100%;
    }
  `}
`;

export const Title = styled.h2`
  margin-bottom: 22px;
  font-size: 24px;
  white-space: nowrap;
  color: ${({ theme }) => theme.colors.text};
`;
