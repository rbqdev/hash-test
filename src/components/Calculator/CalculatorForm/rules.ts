import {
  NAME_AMMOUNT,
  NAME_INSTALLMENTS,
  NAME_MDR
} from "@components/Calculator/constants";

interface IRules {
  [key: string]: (value: number) => boolean;
}

const aMillion = 100000000;

const rules: IRules = {
  [NAME_AMMOUNT]: value => value === 0 || value > aMillion,
  [NAME_INSTALLMENTS]: value => value === 0 || value > 12,
  [NAME_MDR]: value => value === 0 || value > 100
};

export default rules;
