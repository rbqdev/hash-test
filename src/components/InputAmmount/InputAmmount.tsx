import { useRef, useEffect } from "react";
import formatNumber, { currencyConfigs } from "@utils/formatNumber";
import getOnlyDigits from "@utils/getOnlyDigits";
import InputText, { CustomInputProps } from "@components/InputText";

export const LABEL_AMMOUNT = "Informe o valor da venda *";

function InputAmmount({ callbackFunction, name, error }: CustomInputProps) {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  const handleChange = (): void => {
    const element = inputRef?.current;

    if (element === null) {
      return;
    }

    element.value = formatNumber({
      value: element.value,
      ...currencyConfigs
    });

    if (typeof callbackFunction !== "function") {
      return;
    }

    callbackFunction({
      value: getOnlyDigits(element.value),
      name
    });
  };

  return (
    <InputText
      label={LABEL_AMMOUNT}
      aria-label={LABEL_AMMOUNT}
      aria-required="true"
      name={name}
      error={error}
      onChange={handleChange}
      ref={inputRef}
    />
  );
}

export default InputAmmount;
