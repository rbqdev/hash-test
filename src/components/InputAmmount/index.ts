export { default, LABEL_AMMOUNT } from "./InputAmmount";
