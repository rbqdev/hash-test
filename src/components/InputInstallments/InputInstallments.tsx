import { useRef, memo } from "react";
import InputText, { ICustomInput } from "@components/InputText";
import getOnlyDigits from "@utils/getOnlyDigits";

export const LABEL_INSTALLMENTS = "Em quantas parcelas *";

function InputInstallments({ callbackFunction, name, error }: ICustomInput) {
  const inputRef = useRef<HTMLInputElement>(null);

  const handleChange = (): void => {
    const element = inputRef.current;

    if (element === null) {
      return;
    }

    element.value = `${getOnlyDigits(element.value)}`;

    if (typeof callbackFunction !== "function") {
      return;
    }

    callbackFunction({
      value: getOnlyDigits(element.value),
      name
    });
  };

  return (
    <InputText
      label={LABEL_INSTALLMENTS}
      aria-label={LABEL_INSTALLMENTS}
      aria-required="true"
      textInfo="MÃ¡ximo de 12 parcelas"
      name={name}
      error={error}
      onChange={handleChange}
      ref={inputRef}
    />
  );
}

export default memo(InputInstallments);
