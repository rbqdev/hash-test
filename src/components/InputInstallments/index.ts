export { default, LABEL_INSTALLMENTS } from "./InputInstallments";
