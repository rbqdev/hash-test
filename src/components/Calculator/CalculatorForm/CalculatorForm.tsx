import { useState, useContext } from "react";
import { CallbackInputPayload } from "@components/InputText";
import CalculatorContext from "@contexts/Calculator";
import Loader from "@components/Loader";
import InputAmmount from "@components/InputAmmount";
import InputInstallments from "@components/InputInstallments";
import InputMdr from "@components/InputMdr";
import {
  NAME_AMMOUNT,
  NAME_INSTALLMENTS,
  NAME_MDR,
  DEBOUNCE_TIME_FOR_TRIGGER_REQUEST
} from "@components/Calculator/constants";
import useDebounce from "@hooks/useDebounce/useDebounce";
import getListOfDays from "@utils/getListOfDays";
import * as Styled from "./CalculatorForm.styles";
import formRules from "./rules";

interface FormState {
  [key: string]: {
    value: number;
    inputInvalid: boolean;
  };
}

const DeafaultFormValues = {
  value: 0,
  inputInvalid: false
};

function validateInput(name: string, value: number): boolean {
  return formRules[name](value);
}

function validateForm(
  callback: (payload: boolean) => void,
  formValues: FormState
): void {
  const isFormInvalid = Object.keys(formValues).some(key =>
    validateInput(key, formValues[key].value)
  );
  callback(isFormInvalid);
}

export default function CalculatorForm() {
  const [formValues, setFormValues] = useState<FormState>({
    [NAME_AMMOUNT]: DeafaultFormValues,
    [NAME_INSTALLMENTS]: DeafaultFormValues,
    [NAME_MDR]: DeafaultFormValues
  });

  const { isLoading, requestTakeToLong, requestAnticipation } = useContext(
    CalculatorContext
  );

  const handleChangeInput = ({ value, name }: CallbackInputPayload): void => {
    const numberValue = Number(value);

    setFormValues({
      ...formValues,
      [name]: {
        value: numberValue,
        inputInvalid: validateInput(name, numberValue)
      }
    });
  };

  const sendRequest = async (): Promise<void> => {
    const days = getListOfDays();

    await requestAnticipation({
      amount: formValues[NAME_AMMOUNT].value,
      installments: formValues[NAME_INSTALLMENTS].value,
      mdr: formValues[NAME_MDR].value,
      days
    });
  };

  useDebounce(
    () => {
      validateForm(isFormInvalid => {
        if (isFormInvalid) {
          return;
        }

        sendRequest();
      }, formValues);
    },
    DEBOUNCE_TIME_FOR_TRIGGER_REQUEST,
    [formValues]
  );

  return (
    <Styled.Wrapper>
      <Styled.Title>Simule sua Antecipação</Styled.Title>

      <form>
        <InputAmmount
          callbackFunction={handleChangeInput}
          name={NAME_AMMOUNT}
          error={formValues[NAME_AMMOUNT].inputInvalid}
        />
        <InputInstallments
          callbackFunction={handleChangeInput}
          name={NAME_INSTALLMENTS}
          error={formValues[NAME_INSTALLMENTS].inputInvalid}
        />
        <InputMdr
          callbackFunction={handleChangeInput}
          name={NAME_MDR}
          error={formValues[NAME_MDR].inputInvalid}
        />
      </form>

      {isLoading && (
        <Loader
          customText={
            requestTakeToLong
              ? "A requisição ainda está processamento, aguarde..."
              : ""
          }
        />
      )}
    </Styled.Wrapper>
  );
}
