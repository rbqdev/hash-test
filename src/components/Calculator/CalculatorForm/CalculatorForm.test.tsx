import CalculatorContext from "@contexts/Calculator";
import render from "@testsUtils/renderWithTheme";
import { screen, act, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {
  NAME_AMMOUNT,
  NAME_INSTALLMENTS,
  NAME_MDR
} from "@components/Calculator/constants";
import { LABEL_AMMOUNT } from "@components/InputAmmount";
import { LABEL_INSTALLMENTS } from "@components/InputInstallments";
import { LABEL_MDR } from "@components/InputMdr";
import { light as theme } from "@styles/themes";
import getListOfDays from "@utils/getListOfDays";
import CalculatorForm from ".";

const mockRequestAnticipation = jest.fn();

const renderComponent = ({
  response = null,
  error = null,
  isLoading = false,
  requestTakeToLong = false,
  requestAnticipation = mockRequestAnticipation,
  resetFetchValues = jest.fn()
} = {}) =>
  render(
    <CalculatorContext.Provider
      value={{
        response,
        error,
        isLoading,
        requestTakeToLong,
        requestAnticipation,
        resetFetchValues
      }}
    >
      <CalculatorForm />
    </CalculatorContext.Provider>
  );

describe("Component: CalculatorForm", () => {
  it("should render component correctly", () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render title", () => {
    const { getByRole } = renderComponent();
    expect(
      getByRole("heading", { name: /simule sua antecipação/i })
    ).toBeInTheDocument();
  });

  describe("Inputs validations rules", () => {
    const errorStyle = { border: `1px solid ${theme.colors.error}` };

    beforeEach(() => {
      renderComponent();
    });

    test.each([
      ["value === 0", "0"],
      ["value > 100000000", "100000001"]
    ])(
      `should invalidate input ${NAME_AMMOUNT} if value not passaed on rule: '%s'`,
      async (_, value) => {
        const input = screen.getByRole("textbox", {
          name: LABEL_AMMOUNT
        });

        await act(async () => {
          await userEvent.type(input, value);
        });

        expect(input).toHaveStyle(errorStyle);
      }
    );

    test.each([
      ["value === 0", "0"],
      ["value > 12", "13"]
    ])(
      `should invalidate input ${NAME_INSTALLMENTS} if value not passaed on rule: '%s'`,
      async (_, value) => {
        const input = screen.getByRole("textbox", {
          name: LABEL_INSTALLMENTS
        });

        await act(async () => {
          await userEvent.type(input, value);
        });

        expect(input).toHaveStyle(errorStyle);
      }
    );

    test.each([
      ["value === 0", "0"],
      ["value > 100", "101"]
    ])(
      `should invalidate input ${NAME_MDR} if value not passed on rule: '%s'`,
      async (_, value) => {
        const input = screen.getByRole("textbox", {
          name: LABEL_MDR
        });

        await act(async () => {
          await userEvent.type(input, value);
        });

        expect(input).toHaveStyle(errorStyle);
      }
    );
  });

  describe("Event requestAnticipation", () => {
    it("should call if formValues are all valid", async () => {
      jest.useFakeTimers();

      const { getByRole } = renderComponent();

      const inputAmmount = getByRole("textbox", {
        name: LABEL_AMMOUNT
      });
      const inputInstallments = getByRole("textbox", {
        name: LABEL_INSTALLMENTS
      });
      const inputMdr = getByRole("textbox", {
        name: LABEL_MDR
      });

      const days = getListOfDays();

      await act(async () => {
        await userEvent.type(inputAmmount, "2222");
        await userEvent.type(inputInstallments, "12");
        await userEvent.type(inputMdr, "100");

        await waitFor(() =>
          expect(mockRequestAnticipation).toBeCalledWith({
            amount: 2222,
            installments: 12,
            mdr: 100,
            days
          })
        );
      });
    });
  });

  describe("Loader messages", () => {
    const requestLongMessage =
      "A requisição ainda está processamento, aguarde...";

    it("should show loader component when prop 'isLoading' of context are 'true'", async () => {
      const { queryByText } = renderComponent({
        isLoading: true,
        requestTakeToLong: false
      });

      expect(queryByText("Carregando...")).toBeInTheDocument();
      expect(queryByText(requestLongMessage)).not.toBeInTheDocument();
    });

    it("should show message when prop 'requestTakeToLong' of context are 'true'", async () => {
      const { queryByText } = renderComponent({
        isLoading: true,
        requestTakeToLong: true
      });

      expect(queryByText(requestLongMessage)).toBeInTheDocument();
    });
  });
});
