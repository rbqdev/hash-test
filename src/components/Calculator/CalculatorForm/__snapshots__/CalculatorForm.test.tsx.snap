// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CalculatorForm should render component correctly 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin-bottom: 22px;
}

.c2:last-child {
  margin-bottom: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c4 {
  color: #656565;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 14px;
  line-height: 18px;
  margin-bottom: 6px;
}

.c7 {
  color: #cecece;
  font-size: 11px;
  line-height: 14px;
  font-weight: bold;
  margin-top: 4px;
}

.c5 {
  position: relative;
}

.c6 {
  color: #656565;
  background: #fff;
  border: 1px solid #d1dce3;
  border-radius: 4px;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 18px;
  padding: 9px 28px 9px 12px;
  width: 100%;
  outline: none;
}

.c6:hover {
  border: 1px solid #b1c4d0;
}

.c6:focus {
  border: 1px solid #66afe9;
}

.c0 {
  width: 375px;
  padding: 41px 70px 44px 56px;
  border-radius: 4px 0 0 4px;
  background: #fff;
}

.c1 {
  margin-bottom: 22px;
  font-size: 24px;
  white-space: nowrap;
  color: #656565;
}

@media (max-width:768px) {
  .c0 {
    width: 100%;
    height: 100%;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    padding: 25px;
  }

  .c0 form {
    width: 100%;
  }
}

<body>
  <div>
    <section
      class="c0"
    >
      <h2
        class="c1"
      >
        Simule sua Antecipação
      </h2>
      <form>
        <div
          class="c2"
        >
          <label
            class="c3"
            for="ammount"
          >
            <span
              class="c4"
            >
              Informe o valor da venda *
            </span>
            <div
              class="c5"
            >
              <input
                aria-invalid="false"
                aria-label="Informe o valor da venda *"
                aria-required="true"
                class="c6"
                id="ammount"
                name="ammount"
                type="text"
              />
            </div>
            
          </label>
        </div>
        <div
          class="c2"
        >
          <label
            class="c3"
            for="installments"
          >
            <span
              class="c4"
            >
              Em quantas parcelas *
            </span>
            <div
              class="c5"
            >
              <input
                aria-invalid="false"
                aria-label="Em quantas parcelas *"
                aria-required="true"
                class="c6"
                id="installments"
                name="installments"
                type="text"
              />
            </div>
            <small
              class="c7"
            >
              Máximo de 12 parcelas
            </small>
          </label>
        </div>
        <div
          class="c2"
        >
          <label
            class="c3"
            for="mdr"
          >
            <span
              class="c4"
            >
              Informe o percentual de MDR *
            </span>
            <div
              class="c5"
            >
              <input
                aria-invalid="false"
                aria-label="Informe o percentual de MDR *"
                aria-required="true"
                class="c6"
                id="mdr"
                name="mdr"
                type="text"
              />
            </div>
            
          </label>
        </div>
      </form>
    </section>
  </div>
</body>
`;
